# Unstructured PDF Extractuion Pipeline Setup and Testing
This file serves as a playground/workspace to demo Unstructured functionality before it is put into production. This is where we will test the extraction, chunking, metadata extraction, metadata tagging, and vector embedding of PDFs for CASSIE

#### Changelog
**May 18th, 2024**  
**7:00 AM**  
* File created  
* Metadata extraction testing  
**7:30 AM**
* PDF Element Extraction

### Metadata Extraction from PDF Files

#### Imports

In [116]:
import pikepdf
import sys

#### Initial Extraction of Metadata

In [117]:
pdf_file = 'C:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/Security Docs/5000-00-CTSC-48PA-0004_01.pdf'

pdf = pikepdf.Pdf.open(pdf_file)
docInfo = pdf.docinfo

for key, value in docInfo.items():
    print(f'{key}: {value}')

/CreationDate: D:20181004144423-04'00'
/Creator: Microsoft® Word 2013
/ModDate: D:20181004155108-04'00'
/Producer: Microsoft® Word 2013


#### Transform Date Function

In [118]:
import pikepdf
import datetime
import re
from dateutil.tz import tzutc, tzoffset
import sys

In [119]:
pdf_date_pattern = re.compile(''.join([
    r"(D:)?",
    r"(?P<year>/d/d/d/d)",
    r"(?P<month>/d/d)",
    r"(?P<day>/d/d)",
    r"(?P<hour>/d/d)",
    r"(?P<minute>/d/d)",
    r"(?P<second>/d/d)",
    r"(?P<tz_offset>[+-zZ])?",
    r"(?P<tz_hour>/d/d)?",
    r"'?(?P<tz_minute>/d/d)?'?"]))

In [120]:
def transform_date(date_str):
    """
    Convert a pdf date such as "D:20120321183444+07'00'" into a usable datetime
    http://www.verypdf.com/pdfinfoeditor/pdf-date-format.htm
    (D:YYYYMMDDHHmmSSOHH'mm')
    :param date_str: pdf date string
    :return: datetime object
    """
    global pdf_date_pattern
    match = re.match(pdf_date_pattern, date_str)
    if match:
        date_info = match.groupdict()

        for k, v in date_info.items():  # transform values
            if v is None:
                pass
            elif k == 'tz_offset':
                date_info[k] = v.lower()  # so we can treat Z as z
            else:
                date_info[k] = int(v)

        if date_info['tz_offset'] in ('z', None):  # UTC
            date_info['tzinfo'] = tzutc()
        else:
            multiplier = 1 if date_info['tz_offset'] == '+' else -1
            date_info['tzinfo'] = tzoffset(None, multiplier*(3600 * date_info['tz_hour'] + 60 * date_info['tz_minute']))

        for k in ('tz_offset', 'tz_hour', 'tz_minute'):  # no longer needed
            del date_info[k]

        return datetime.datetime(**date_info)

#### Extract Dates from PDF

In [121]:
pdf = pikepdf.Pdf.open(pdf_file)
docinfo = pdf.docinfo
for key, value in docinfo.items():
    if str(value).startswith("D:"):
        # pdf datetime format, convert to python datetime
        value = transform_date(str(pdf.docinfo["/CreationDate"]))
    print(key, ":", value)

/CreationDate : None
/Creator : Microsoft® Word 2013
/ModDate : None
/Producer : Microsoft® Word 2013


So it looks like the created and modified dates can be extracted from PDFs, even though I've downloaded them much later. I will try this for all the PDF files in the CSDocs folder to ensure it works properly

In [122]:
import os

In [123]:
CSDocs_path = 'C:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/'

doc_info = []


for file in os.listdir(CSDocs_path):
        try:
            doc = {}
            if file.endswith('.pdf'):
                doc = {'filename': file}
                pdf = pikepdf.Pdf.open(CSDocs_path + file)
                docInfo = pdf.docinfo
                for key, value in docInfo.items():
                    print(f'{key}: {value}')
                    if str(value).startswith("D:"):
                        # pdf datetime format, convert to python datetime
                        value = transform_date(str(pdf.docinfo["/CreationDate"]))
                    if key == '/CreationDate':
                        key = 'Creation Date'
                    if key == '/Creator':
                        key = 'Creator'
                    if key == '/ModDate':
                        key = 'Modification Date'
                    if key == '/Producer':
                        key = 'Producer'
                    if key == '/Title':
                        key = 'Title'
                    if key == '/Subject':
                        key = 'Subject'
                    if key == '/Author':
                        key = 'Author'
                    doc[key] = value
                print('\n')
            if doc['filename']:
                doc_info.append(doc)
        except Exception as e:
            print(e)

/Author: Yousef Kimiagar
/CreationDate: D:20200210124719-05'00'
/Creator: Microsoft® Word for Office 365
/ModDate: D:20200210124719-05'00'
/Producer: Microsoft® Word for Office 365
/Title: SCHEDULE 15-2


/Author: David Robson/Duncan Robb
/CreationDate: D:20170927074532-04'00'
/Creator: Microsoft® Word 2013
/ModDate: D:20170927083115-04'00'
/Producer: Microsoft® Word 2013
/Subject: Eglinton Crosstown LRT Project
/Title: ECLRT Cyber Security Management Plan


'filename'
/Author: 
/CreationDate: D:20220506210237-04'00'
/Creator: Aspose Ltd.
/ModDate: D:20220506210237-04'00'
/Producer: Aspose.Pdf for .NET 11.4.0
/Subject: 
/Title: 


/CreationDate: D:20210831101215-04'00'
/Creator: Microsoft® Word for Microsoft 365
/ModDate: D:20220224143709-05'00'
/Producer: Microsoft® Word for Microsoft 365


/CreationDate: D:20210714131927-04'00'
/Creator: Microsoft® Word for Microsoft 365
/ModDate: D:20220224145956-05'00'
/Producer: Microsoft® Word for Microsoft 365


/CreationDate: D:20210715181444-0

In [124]:
for i in doc_info:
    for key, value in i.items():
        print(f'{key}: {value}')
    print('\n')



filename: 15-2 Conformed.pdf
Author: Yousef Kimiagar
Creation Date: None
Creator: Microsoft® Word for Office 365
Modification Date: None
Producer: Microsoft® Word for Office 365
Title: SCHEDULE 15-2


filename: 5000-00-CTSC-48PA-0002 - Cyber Security Management Plan-5000-00-CTSC-48PA-0002_PA.pdf
Author: David Robson/Duncan Robb
Creation Date: None
Creator: Microsoft® Word 2013
Modification Date: None
Producer: Microsoft® Word 2013
Subject: Eglinton Crosstown LRT Project
Title: ECLRT Cyber Security Management Plan


filename: 5000-00-WGD-48PA- 2001 - Internal Handover Plan – Cyber Systems-2700-67ASA1-02-CTSC-0004_01.pdf
Author: 
Creation Date: None
Creator: Aspose Ltd.
Modification Date: None
Producer: Aspose.Pdf for .NET 11.4.0
Subject: 
Title: 


filename: 5000-00-WGD-48PA-1001 - Baseline Controls Document - Management Controls Policy.pdf
Creation Date: None
Creator: Microsoft® Word for Microsoft 365
Modification Date: None
Producer: Microsoft® Word for Microsoft 365


filename: 5000-

In [125]:
for i in doc_info:
    print(i)

{'filename': '15-2 Conformed.pdf', 'Author': pikepdf.String("Yousef Kimiagar"), 'Creation Date': None, 'Creator': pikepdf.String("Microsoft® Word for Office 365"), 'Modification Date': None, 'Producer': pikepdf.String("Microsoft® Word for Office 365"), 'Title': pikepdf.String("SCHEDULE 15-2")}
{'filename': '5000-00-CTSC-48PA-0002 - Cyber Security Management Plan-5000-00-CTSC-48PA-0002_PA.pdf', 'Author': pikepdf.String("David Robson/Duncan Robb"), 'Creation Date': None, 'Creator': pikepdf.String("Microsoft® Word 2013"), 'Modification Date': None, 'Producer': pikepdf.String("Microsoft® Word 2013"), 'Subject': pikepdf.String("Eglinton Crosstown LRT Project"), 'Title': pikepdf.String("ECLRT Cyber Security Management Plan")}
{'filename': '5000-00-WGD-48PA- 2001 - Internal Handover Plan – Cyber Systems-2700-67ASA1-02-CTSC-0004_01.pdf', 'Author': pikepdf.String(""), 'Creation Date': None, 'Creator': pikepdf.String("Aspose Ltd."), 'Modification Date': None, 'Producer': pikepdf.String("Aspose.P

So it looks like there is a good amount of metadata already stored in the metadata of the PDF. We will aim to enhance this metadata using PDF extraction and element recognition. We'll start by reading a PDF with unstructured and seeing what elements exist on the first page/author block.

It is notable that the title metadata isn't always descriptive. I'm not sure how we can get a document description using an LLM, but I think either RAPTOR (Continuously shrinking documents (Subsection Summaries => Section Summaries => Doc Summaries) could work to shorten the tokens needed, or passing to Gemini (The 1M, soon 2M, token context window should be enough to summarize most documents) could be a good solution.)

We will start with Pulling Titles and Authors which starts with Unstructured extraction

### Unstuctured Extraction

In [126]:
pdf_file = "c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf"

from unstructured.partition.pdf import partition_pdf

elements = partition_pdf(
    pdf_file,
    languages=['en'],
    hi_res_model_name='yolox',
    infer_table_structure=True,
    strategy='hi_res'
    )

In [127]:
print(len(elements))
metadata = elements[0].metadata.to_dict()
el = elements[1].to_dict()

for element in elements:
    element_dict = element.to_dict()
    if element_dict['type'] == 'Title':
        print(element_dict['text'])
        print('\n')
        first_title_element = element_dict
        break

for key, value in metadata.items():
    print(f'{key}: {value}')

print('\n')
for key, value in first_title_element.items():
    print(f'{key}: {value}')

504
Cyber Security – Change Control and Configuration Management Process


coordinates: {'points': ((192.29166666666666, 310.55555555555543), (192.29166666666666, 532.3611111111111), (749.2361111111112, 532.3611111111111), (749.2361111111112, 310.55555555555543)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200}
last_modified: 2023-03-28T16:34:32
filetype: application/pdf
languages: ['en']
page_number: 1
file_directory: c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs
filename: 5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf


type: Title
element_id: 8df1b47d06638e6081109dc18f6b42d0
text: Cyber Security – Change Control and Configuration Management Process
metadata: {'detection_class_prob': 0.48540282249450684, 'coordinates': {'points': ((322.7745666503906, 599.533203125), (322.7745666503906, 751.4722222222222), (1370.8916015625, 751.4722222222222), (1370.8916015625, 599.533203125)), 'system': 'PixelSpace', 'layout_width': 1700, 'layo

The file has been partitioned using the yolox model. 
filename: 5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf  
elements: 504  
first title element text: Cyber Security – Change Control and Configuration Management Process  
metadata:  
    **coordinates**:   
        {'points': ((192.29166666666666, 310.55555555555543), (192.29166666666666, 532.3611111111111), (749.2361111111112, 532.3611111111111), (749.2361111111112, 310.55555555555543)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200}
    **last_modified**:   
        2023-03-28T16:34:32
    **filetype**:   
        application/pdf
    **languages**:   
        ['en']
    **page_number**:   
        1
    **file_directory**:  
        c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs
    **filename**:   
        5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf

#### Examining Headers and Footers

In [128]:
for i in elements:
    print(i.to_dict()['type'])
    if i.to_dict()['type'] == 'Table':
        print(i.metadata.text_as_html)
    elif i.to_dict()['type'] == 'NarrativeText':
        print(i.to_dict()['text'])
    elif i.to_dict()['type'] == 'Title':
        print(i.to_dict()['text'])
    elif i.to_dict()['type'] == 'ListItem':
        print(i.to_dict()['text'])
    elif i.to_dict()['type'] == 'UncategorizedText':
        print(i.to_dict()['text'])
    else:
        print(i.to_dict()['type'])

Image
Image
Image
Image
Title
Cyber Security – Change Control and Configuration Management Process
Image
Image
Image
Image
Image
Image
NarrativeText
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. Please access the electronic version for the most current information.
Image
Image
UncategorizedText
5000-00-WGD-48PA-1020
UncategorizedText
03
Title
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Title
Doc Owner
NarrativeText
WGD Consulting
Title
DOCUMENT REVISIONS INDEX
Table
<table><thead><th>Revision</th><th>Revision Date</th><th>Author(s)</th><th>Reviewer(s)</th><th>Description of Changes</th></thead><tr><td>PA</td><td>April 8, 2020</td><td>Chris Goulopou

In [129]:
headers_list = [el for el in elements if el.to_dict()['type'] == 'Header']
footers_list = [el for el in elements if el.to_dict()['type'] == 'Footer']

In [130]:
print(f"Number of Headers: {len(headers_list)}")
print(f"Number of Footers: {len(footers_list)}")

Number of Headers: 6
Number of Footers: 4


In [131]:
for i in headers_list:
    print(i.to_dict()['text'])
for i in footers_list:
    print(i.to_dict()['text'])

Doc Owner
Doc Owner
WGD Consulting
Doc Owner
WGD Consulting
Doc Owner
Page | 22 of 25
Document Classification: CTSC/ECLRT Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. Please access the electronic version for the most current information.
Page | 23 of 25
Page | 24 of 25


In [132]:
tables_list = [el for el in elements if el.to_dict()['type'] == 'Table']

for i in tables_list:
    print(i.metadata.text_as_html)

<table><thead><th>Revision</th><th>Revision Date</th><th>Author(s)</th><th>Reviewer(s)</th><th>Description of Changes</th></thead><tr><td>PA</td><td>April 8, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Initial draft for review</td></tr><tr><td>00</td><td>April 24, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Updated to reflect comments from review</td></tr><tr><td>01</td><td>July 9, 2020</td><td>Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td></tr><tr><td>02</td><td>August 27, 2020</td><td>Mark Salsberg</td><td>Chris Goulopoulos</td><td>Minor corrections and edits</td></tr><tr><td>03</td><td>June 29, 2021</td><td>. Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Adjusted wording for responsibilities based on CTS contractual agreements</td></tr></table>
<table><tr><td>PULDOSEC</td><td>.... ee</td><td>eeeeecceecceeceeeces

In [46]:
def count_elements(elements):
    element_count = {}
    for element in elements:
        element_type = element.category
        if element_type not in element_count:
            element_count[element_type] = 0
        element_count[element_type] += 1
    return element_count

In [134]:
element_count = count_elements(elements)
for i in element_count:
    print(f'{i}: {element_count[i]}')

Image: 37
Title: 104
NarrativeText: 139
Text: 59
Table: 9
ListItem: 146
Header: 6
Footer: 4


In [135]:
titles = [el for el in elements if el.to_dict()['type'] == 'Title']
for i in titles:
    print(i.to_dict()['text'])

Cyber Security – Change Control and Configuration Management Process
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
DOCUMENT REVISIONS INDEX
REVISION AND CONTROL
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
Contents
Appendix A
Appendix B
Appendix C
Appendix D
Appendix E
Appendix F
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
1 Introduction
2 Purpose
3 Objectives
4 Scope and Applicability
Cyber Security – Change Control and Configuration Management Process 2021-03-01
4.1 Applicability
4.2 Exclusions
4.3 Associated Documents
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
5 Roles and Responsibilities
1. Cyber Security Senior Manager
2. Delegates (Functional Delegates)
3. Technical and Compliance Oversight
Change Review Group
4.

In [136]:
from io import StringIO 
from lxml import etree

table_html = tables_list[0].metadata.text_as_html

parser = etree.XMLParser(remove_blank_text=True)
file_obj = StringIO(table_html)
tree = etree.parse(file_obj, parser)
print(etree.tostring(tree, pretty_print=True).decode())

<table>
  <thead>
    <th>Revision</th>
    <th>Revision Date</th>
    <th>Author(s)</th>
    <th>Reviewer(s)</th>
    <th>Description of Changes</th>
  </thead>
  <tr>
    <td>PA</td>
    <td>April 8, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Initial draft for review</td>
  </tr>
  <tr>
    <td>00</td>
    <td>April 24, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Updated to reflect comments from review</td>
  </tr>
  <tr>
    <td>01</td>
    <td>July 9, 2020</td>
    <td>Chris Goulopoulos_</td>
    <td>| Mark Salsberg</td>
    <td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td>
  </tr>
  <tr>
    <td>02</td>
    <td>August 27, 2020</td>
    <td>Mark Salsberg</td>
    <td>Chris Goulopoulos</td>
    <td>Minor corrections and edits</td>
  </tr>
  <tr>
    <td>03</td>
    <td>June 29, 2021</td>
    <td>. Chris Goulopoulos_</td>
    <td>| Mark

In [137]:
from IPython.core.display import HTML
HTML(table_html)

In [138]:
from langchain_openai import ChatOpenAI
from langchain_core.documents import Document
from langchain.chains.summarize import load_summarize_chain

In [139]:
openai_api_key = os.environ.get('OPENAI_API_KEY')

In [140]:
os.environ['OPENAI_API_KEY'] = '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'
llm = ChatOpenAI(temperature=0,openai_api_key=openai_api_key, model_name="gpt-4o")
chain = load_summarize_chain(llm, chain_type="stuff")
chain.invoke([Document(page_content=table_html)])

{'input_documents': [Document(page_content='<table><thead><th>Revision</th><th>Revision Date</th><th>Author(s)</th><th>Reviewer(s)</th><th>Description of Changes</th></thead><tr><td>PA</td><td>April 8, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Initial draft for review</td></tr><tr><td>00</td><td>April 24, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Updated to reflect comments from review</td></tr><tr><td>01</td><td>July 9, 2020</td><td>Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td></tr><tr><td>02</td><td>August 27, 2020</td><td>Mark Salsberg</td><td>Chris Goulopoulos</td><td>Minor corrections and edits</td></tr><tr><td>03</td><td>June 29, 2021</td><td>. Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Adjusted wording for responsibilities based on CTS contractual agreements</td></tr></table>')],
 'output_text': 'The

In [ ]:
from unstructured.chunking.title import chunk_by_title

In [158]:
images = [el for el in elements if el.to_dict()['type'] == 'Image']

for i in images:
    print(i)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
  
Baseline Configuration Struc
Cyber Asset Baseline Configuration List |Review/Update Date: aor Document Classification Level: Cybersecurity Information - Confidential ‘Configuration Change st = pom i a — A cert eta pan EM Pin et ug “ 23 _|LAN Switch [Network switch '- Network Switches 24_|LAN Switch 

In [152]:
doc_text = '' 
for el in elements:
    if el.to_dict()['type'] != 'Header' and el.to_dict()['type'] != 'Footer':
        if el.to_dict()['type'] == 'Table':
            table_html = el.metadata.text_as_html

            parser = etree.XMLParser(remove_blank_text=True)
            file_obj = StringIO(table_html)
            tree = etree.parse(file_obj, parser)
            print(etree.tostring(tree, pretty_print=True).decode())
            doc_text += etree.tostring(tree, pretty_print=True).decode() + '\n'
        else:
            doc_text += el.to_dict()['text'] + '\n'

<table>
  <thead>
    <th>Revision</th>
    <th>Revision Date</th>
    <th>Author(s)</th>
    <th>Reviewer(s)</th>
    <th>Description of Changes</th>
  </thead>
  <tr>
    <td>PA</td>
    <td>April 8, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Initial draft for review</td>
  </tr>
  <tr>
    <td>00</td>
    <td>April 24, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Updated to reflect comments from review</td>
  </tr>
  <tr>
    <td>01</td>
    <td>July 9, 2020</td>
    <td>Chris Goulopoulos_</td>
    <td>| Mark Salsberg</td>
    <td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td>
  </tr>
  <tr>
    <td>02</td>
    <td>August 27, 2020</td>
    <td>Mark Salsberg</td>
    <td>Chris Goulopoulos</td>
    <td>Minor corrections and edits</td>
  </tr>
  <tr>
    <td>03</td>
    <td>June 29, 2021</td>
    <td>. Chris Goulopoulos_</td>
    <td>| Mark

In [153]:
print(doc_text)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. 

In [154]:
chatInstance = ChatOpenAI(temperature=0, openai_api_key=openai_api_key, model_name="gpt-4o")
chatInstance.get_num_tokens(doc_text)

14452

In [155]:
response = chatInstance.invoke(f"Clean the following document to be embedded in a vector database. Return only the cleaned document text: {doc_text}")

In [157]:
print(response.content)

Crosslinx Transit Solutions (CTS) has implemented a Cyber Security Plan (CSP) and associated policies and processes for protecting the Eglinton Crosstown Light Rail Transit (ECLRT’s) Cyber Systems from cyber security threats that could compromise the safe and reliable operation of the ECLRT.

The CTS Cyber Security - Management Controls Policy, establishes the management controls and the framework for cyber security standards, plans and processes for securing those Cyber Systems that are essential for the safe and reliable operation of the ECLRT.

The Cyber Security – Change Control and Configuration Management Standard, defined the cyber security controls and required processes for managing changes to ECLRT’s operational Cyber Systems and associated Cyber Assets and their configurations.

This document presents the processes for implementing, sustaining, and adhering to the cyber security controls for managing and controlling changes to ECLRT’s Cyber Systems and changes to the baselin

#### Conclusions
**Good Information**  
It seems like we can extract all the information with a good deal of clarity. Combining Unstructured with GPT-4o could prove incredibly useful. Basically, we would extract all the chunks, and then, for all chunks that arent headers or footers, we would pass the elements text property (text_as_html property for tables) to GPT-4o to clean the text for text elements, and summarize the tables for table elements. We would then concatenate all the information into one document, then pass that to GPT-4o to pull Title, Doc type, section headings, topics, authors, revision dates, 

In [159]:
response_2 = chatInstance.invoke(f"Extract the document title, along with the section headers from the following document: {doc_text}")

In [160]:
print(response_2.content)

**Document Title:**
Cyber Security – Change Control and Configuration Management Process

**Section Headers:**
1. Introduction
2. Purpose
3. Objectives
4. Scope and Applicability
   - 4.1 Applicability
   - 4.2 Exclusions
   - 4.3 Associated Documents
5. Roles and Responsibilities
6. Definitions and Acronyms
7. Change and Configuration Management Overview
8. Change Control and Configuration Management Process
   - 8.1 Configuration Baseline Structure/Attributes
   - 8.2 Establishing the Configuration Baseline
     - 8.2.1 Test and Commissioning/As Build - Configuration Baseline
     - 8.2.2 Production/Commercial Operations - Configuration Baseline
     - 8.2.3 Configuration Management Tools
   - 8.3 Guidelines for Establishing the Baseline Configuration
     - 8.3.1 Pre-Commissioning - Scanning for Configurations
     - 8.3.2 Production - Scanning Cyber Asset Configurations
       - 8.3.2.1 Non-Production Environment Configuration Scanning Method
       - 8.3.2.2 Production Environment

In [163]:
chat_instance_json = ChatOpenAI(temperature=0, openai_api_key=openai_api_key, model_name="gpt-4o",response_format={ "type": "json_object" })

c:\Users\Dan\AppData\Local\Programs\Python\Python311\Lib\site-packages\langchain_core\utils\utils.py:159: UserWarning: WARNING! response_format is not default parameter.
                response_format was transferred to model_kwargs.
                Please confirm that response_format is what you intended.
  warnings.warn(


In [173]:
response_2 = chat_instance_json.invoke(f"Extract the document title (as document_title), along with the section headers (as section_headers) from the following document along with document type (as document_type) (procedure, technical document, system specification, project agreement, contract, request for proposal (RFP), bid, etc.) and all the tags of the document (as tags) (Use your judgement for this. The tags could be (the tags don't have to be in this list): cybersecurity, ECLRT, change management, configuration management, *System names*, schematic, process, architecture document, As-Built, etc.) and the authors (as authors) of the document and the revision number (as revision). Return them in a JSON object: {doc_text}", response_format={ "type": "json_object" })

In [195]:
print(response_2.content)

{
  "document_title": "Cyber Security – Change Control and Configuration Management Process",
  "section_headers": [
    "Introduction",
    "Purpose",
    "Objectives",
    "Scope and Applicability",
    "Applicability",
    "Exclusions",
    "Associated Documents",
    "Roles and Responsibilities",
    "Definitions and Acronyms",
    "Change and Configuration Management Overview",
    "Change Control and Configuration Management Process",
    "Configuration Baseline Structure/Attributes",
    "Establishing the Configuration Baseline",
    "Test and Commissioning/As Build - Configuration Baseline",
    "Production/Commercial Operations - Configuration Baseline",
    "Configuration Management Tools",
    "Guidelines for Establishing the Baseline Configuration",
    "Pre-Commissioning - Scanning for Configurations",
    "Production - Scanning Cyber Asset Configurations",
    "Non-Production Environment Configuration Scanning Method",
    "Production Environment Configuration Scanning Me

### Process Flow

1. Extract doc elements
2. Remove headers and footers
3. Pass to LLM  
    a. summarize if table, extract topics in json mode  
    b. clean if else, extract topics in json mode
4. Concatenate elements
5. Extract doc metadata
6. Assign section titles to elements
7. Assign doc_wide metadata to all elements
8. Embed and store

In [1]:
## Extract PDF elements

from unstructured.partition.pdf import partition_pdf

pdf_file = "c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf"

elements = partition_pdf(
    pdf_file,
    languages=['en'],
    hi_res_model_name='yolox',
    infer_table_structure=True,
    strategy='hi_res'
    )

Some weights of the model checkpoint at microsoft/table-transformer-structure-recognition were not used when initializing TableTransformerForObjectDetection: ['model.backbone.conv_encoder.model.layer2.0.downsample.1.num_batches_tracked', 'model.backbone.conv_encoder.model.layer3.0.downsample.1.num_batches_tracked', 'model.backbone.conv_encoder.model.layer4.0.downsample.1.num_batches_tracked']
- This IS expected if you are initializing TableTransformerForObjectDetection from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TableTransformerForObjectDetection from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [2]:
for v in elements:
    print(v)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. 

In [18]:
## Remove Header and Footer elements

def remove_headers_and_footers(elements):
    cleaned_elements = [element for element in elements if element.to_dict()['type'] not in ['Header', 'Footer']]
    return cleaned_elements

# Test header/footer removal
print(len(elements))  # Display the number of elements before cleaning

# Create a copy of the elements list to avoid modifying the original list
import copy

cleaned_elements = copy.deepcopy(remove_headers_and_footers(elements))
print(f"Number of elements after cleaning: {len(cleaned_elements)}")  # Display the number of elements after cleaning

504
Number of elements after cleaning: 494


In [16]:
## LLM Cleaning and Topic Extraction
import json
from langchain_openai import ChatOpenAI
import os
os.environ['OPENAI_API_KEY'] = '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'
openai_api_key = os.environ.get('OPENAI_API_KEY')
OpenCleaner = ChatOpenAI(temperature=0, openai_api_key=openai_api_key, model_name="gpt-4o")

for element in cleaned_elements:
    if element.category == 'Table':
        prompt = f"Summarize the following table in detail and return only the text summary with no extra characters:\n\n{element.metadata.text_as_html}"
    else:
        prompt = f"Clean and return the following text. Do not modify the content of the text or add any information, just remove erroneous characters, and adjust the formatting. If there is no text, return 'N/A':\n\n{element.text}"

    response = OpenCleaner.invoke(prompt)
    print(element.text)
    element.text = response.content


CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. 

In [17]:
for element in cleaned_elements:
    print(element.text)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021
Cyber Security – Change Control and Configuration Management Process
Prepared by:   
Reviewed by:   
Reviewed by:   
Approved by:   
Accepted by:   
For Eglinton Crosstown LRT   
Chris Goulopoulos  
Consultant, CTS   
William VanRyswyk, CPP  
Chief Security Officer, CTS   
Evan Harben  
Cyber Security & OT Network Specialist, CTSC  
Hadi Meshgi  
Technical Lead - Communication & Cybersecurity, CTSC   
Oct 20, 2021  
Kyle Booth  
Systems Integration Authority, CTSC   
Name, Title   
Signature/Date   
Document No. 5000-00-WGD-48PA-1020   
Rev. 03   
July 26, 2021
if I A vo hay Gi x 4 July Lp
Nee gle "dr"
© 2021 Crosslinx Transit Solutions

This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrol

In [12]:
## Concatenate cleaned elements into a single document text

doc_text = ''
for element in cleaned_elements:
    doc_text += element.text + '\n'

print(doc_text)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. 

In [72]:
from io import StringIO 
from lxml import etree

doc_text = '' 
for el in cleaned_elements:
    if el.to_dict()['type'] != 'Header' and el.to_dict()['type'] != 'Footer':
        if el.category == 'Table':
            table_html = el.metadata.text_as_html

            parser = etree.XMLParser(remove_blank_text=True)
            file_obj = StringIO(table_html)
            tree = etree.parse(file_obj, parser)
            # print(etree.tostring(tree, pretty_print=True).decode())
            el.html = etree.tostring(tree, pretty_print=True).decode()
            print(el.html)
            doc_text += etree.tostring(tree, pretty_print=True).decode() + '\n'
        else:
            doc_text += el.to_dict()['text'] + '\n'


<table>
  <thead>
    <th>Revision</th>
    <th>Revision Date</th>
    <th>Author(s)</th>
    <th>Reviewer(s)</th>
    <th>Description of Changes</th>
  </thead>
  <tr>
    <td>PA</td>
    <td>April 8, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Initial draft for review</td>
  </tr>
  <tr>
    <td>00</td>
    <td>April 24, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Updated to reflect comments from review</td>
  </tr>
  <tr>
    <td>01</td>
    <td>July 9, 2020</td>
    <td>Chris Goulopoulos_</td>
    <td>| Mark Salsberg</td>
    <td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td>
  </tr>
  <tr>
    <td>02</td>
    <td>August 27, 2020</td>
    <td>Mark Salsberg</td>
    <td>Chris Goulopoulos</td>
    <td>Minor corrections and edits</td>
  </tr>
  <tr>
    <td>03</td>
    <td>June 29, 2021</td>
    <td>. Chris Goulopoulos_</td>
    <td>| Mark

In [25]:
print(doc_text)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. 

In [26]:
prompt = f"Extract the document title (as document_title), the document authors (as authors), the date of publication (as date), the revision number (as revision_number), and all section headings (as section_headings). Return all of these in JSON format:\n\n{doc_text}"

doc_metadata = OpenCleaner.invoke(prompt, response_format={ "type": "json_object" })

In [27]:
print(doc_metadata.content)

{
  "document_title": "Cyber Security – Change Control and Configuration Management Process",
  "authors": [
    "Chris Goulopoulos",
    "William VanRyswyk, CPP",
    "Evan Harben",
    "Hadi Meshgi",
    "Kyle Booth"
  ],
  "date": "2021-10-20",
  "revision_number": "03",
  "section_headings": [
    "1 Introduction",
    "2 Purpose",
    "3 Objectives",
    "4 Scope and Applicability",
    "4.1 Applicability",
    "4.2 Exclusions",
    "4.3 Associated Documents",
    "5 Roles and Responsibilities",
    "6 Definitions and Acronyms",
    "7 Change and Configuration Management Overview",
    "8 Change Control and Configuration Management Process",
    "8.1 Configuration Baseline Structure/Attributes",
    "8.2 Establishing the Configuration Baseline",
    "8.2.1 Test and Commissioning/As Build - Configuration Baseline",
    "8.2.2 Production/Commercial Operations - Configuration Baseline",
    "8.2.3 Configuration Management Tools",
    "8.3 Guidelines for Establishing the Baseline Conf

In [28]:
doc_metadata_dict = json.loads(doc_metadata.content)
print(doc_metadata_dict.keys())

dict_keys(['document_title', 'authors', 'date', 'revision_number', 'section_headings'])


In [29]:
section_headings = doc_metadata_dict['section_headings']

print(section_headings)

['1 Introduction', '2 Purpose', '3 Objectives', '4 Scope and Applicability', '4.1 Applicability', '4.2 Exclusions', '4.3 Associated Documents', '5 Roles and Responsibilities', '6 Definitions and Acronyms', '7 Change and Configuration Management Overview', '8 Change Control and Configuration Management Process', '8.1 Configuration Baseline Structure/Attributes', '8.2 Establishing the Configuration Baseline', '8.2.1 Test and Commissioning/As Build - Configuration Baseline', '8.2.2 Production/Commercial Operations - Configuration Baseline', '8.2.3 Configuration Management Tools', '8.3 Guidelines for Establishing the Baseline Configuration', '8.3.1 Pre-Commissioning - Scanning for Configurations', '8.3.2 Production - Scanning Cyber Asset Configurations', '8.3.2.1 Non-Production Environment Configuration Scanning Method', '8.3.2.2 Production Environment Configuration Scanning Method', '8.4 Change Control Management Process', '8.4.1 Planned/Scheduled - Change Management', '8.4.1.1 Preparing 

In [73]:
test_cleaned_elements = copy.deepcopy(cleaned_elements)

In [74]:
print(f"Number of Title elements before reclassification: {len([element for element in test_cleaned_elements if element.category == 'Title'])}")

count = 0
for element in test_cleaned_elements:
    if element.category == 'Title':
        if element.text not in section_headings:
            element.category = 'NarrativeText'
            print(element.category)
    else:
        print('Title Found')

print(f"Number of Title elements after reclassification: {len([element for element in test_cleaned_elements if element.category == 'Title'])}")
print(f"Number of Title elements after reclassification: {len([element for element in cleaned_elements if element.category == 'Title'])}")

Number of Title elements before reclassification: 104
Title Found
Title Found
NarrativeText
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
NarrativeText
NarrativeText
Title Found
NarrativeText
Title Found
NarrativeText
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
NarrativeText
NarrativeText
Title Found
NarrativeText
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
NarrativeText
NarrativeText
NarrativeText
NarrativeText
NarrativeText
NarrativeText
Title Found
Title Found
Title Found
Title Found
Title Found
NarrativeText
NarrativeText
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
NarrativeText
NarrativeText
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Found
Title Foun

In [75]:
element_count = count_elements(test_cleaned_elements)
print(element_count)
print(len(section_headings))


{'Image': 37, 'NarrativeText': 205, 'UncategorizedText': 59, 'Table': 9, 'Title': 38, 'ListItem': 146}
40


In [76]:
for element in test_cleaned_elements:
    if element.category == 'Title':
        print(element.to_dict())

{'type': 'Title', 'element_id': 'd0fe462b5190c2d5c762e682a4a25069', 'text': '1 Introduction', 'metadata': {'detection_class_prob': 0.8250165581703186, 'coordinates': {'points': ((199.08587646484375, 238.45797729492188), (199.08587646484375, 292.22222222222234), (531.3555555555556, 292.22222222222234), (531.3555555555556, 238.45797729492188)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200}, 'last_modified': '2023-03-28T16:34:32', 'filetype': 'application/pdf', 'languages': ['en'], 'page_number': 5, 'file_directory': 'c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs', 'filename': '5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf'}}
{'type': 'Title', 'element_id': '4241c7db56ebac1441c8102176dd69e0', 'text': '2 Purpose', 'metadata': {'coordinates': {'points': ((200.06666666666666, 918.6388888888888), (200.06666666666666, 970.6388888888889), (452.6888888888889, 970.6388888888889), (452.6888888888889, 918.6388888888888)), 'system': 'PixelS

In [228]:
chunks = chunk_by_title(test_cleaned_elements)

In [230]:
print(len(chunks))

161


In [229]:
for chunk in chunks:
    print(chunk)
    print("\n\n" + "-"*80)
    input()
    

CROSSLIN TRANSIT SOLUTIONS

Oct 20, 2021


--------------------------------------------------------------------------------
Cyber Security – Change Control and Configuration Management Process

Prepared by: Chris Goulopoulos Consultant, CTS
Reviewed by: William VanRyswyk, CPP Chief Security Officer, CTS
Reviewed by: Evan Harben Cyber Security & OT Network Specialist, CTSC
Approved by: Hadi Meshgi Technical Lead - Communication & Cybersecurity, CTSC Oct 20, 2021
Accepted by: Kyle Booth Systems Integration Authority, CTSC
For Eglinton Crosstown LRT
Document No. 5000-00-WGD-48PA-1020
Rev. 03
July 26, 2021


--------------------------------------------------------------------------------
if iA vo hay Gi x 4 July Lp

Nee gle “dr

This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. Please access the electronic version for the most current information.

Sure, please provide the te

In [67]:
element_count = count_elements(test_cleaned_elements)
for i in element_count:
    print(f'{i}: {element_count[i]}')

Image: 37
NarrativeText: 205
UncategorizedText: 59
Table: 9
Title: 38
ListItem: 146


In [49]:
from unstructured.chunking.title import chunk_by_title

chunks = chunk_by_title(test_cleaned_elements)

In [68]:
element_count = count_elements(test_cleaned_elements)
for i in element_count:
    print(f'{i}: {element_count[i]}')

Image: 37
NarrativeText: 205
UncategorizedText: 59
Table: 9
Title: 38
ListItem: 146


In [77]:
for element in test_cleaned_elements:
    print(element.to_dict())

{'type': 'Image', 'element_id': 'b13a99fcb24c0a177e73cb6913c9a38f', 'text': 'CROSSLIN TRANSIT SOLUTIONS', 'metadata': {'coordinates': {'points': ((192.29166666666666, 310.55555555555543), (192.29166666666666, 532.3611111111111), (749.2361111111112, 532.3611111111111), (749.2361111111112, 310.55555555555543)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200}, 'last_modified': '2023-03-28T16:34:32', 'filetype': 'application/pdf', 'languages': ['en'], 'page_number': 1, 'file_directory': 'c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs', 'filename': '5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf'}}
{'type': 'Image', 'element_id': 'ca023571f9c731558cc15b2508fc166d', 'text': 'Oct 20, 2021 ', 'metadata': {'coordinates': {'points': ((966.8611111111111, 513.3333333333331), (966.8611111111111, 1660.575), (2477.241666666667, 1660.575), (2477.241666666667, 513.3333333333331)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200

In [78]:
for element in test_cleaned_elements:
    if element.category == 'Table':
        print(element.html)

<table>
  <thead>
    <th>Revision</th>
    <th>Revision Date</th>
    <th>Author(s)</th>
    <th>Reviewer(s)</th>
    <th>Description of Changes</th>
  </thead>
  <tr>
    <td>PA</td>
    <td>April 8, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Initial draft for review</td>
  </tr>
  <tr>
    <td>00</td>
    <td>April 24, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Updated to reflect comments from review</td>
  </tr>
  <tr>
    <td>01</td>
    <td>July 9, 2020</td>
    <td>Chris Goulopoulos_</td>
    <td>| Mark Salsberg</td>
    <td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td>
  </tr>
  <tr>
    <td>02</td>
    <td>August 27, 2020</td>
    <td>Mark Salsberg</td>
    <td>Chris Goulopoulos</td>
    <td>Minor corrections and edits</td>
  </tr>
  <tr>
    <td>03</td>
    <td>June 29, 2021</td>
    <td>. Chris Goulopoulos_</td>
    <td>| Mark

In [94]:
def chunk_by_previous_title(elements):
    chunks=[]
    current_chunk = ""
    current_title = None

    for element in elements:
        if element.category == 'Title':
            if current_chunk:
                chunks.append(current_chunk)
                current_chunk = ""
            current_title = element.text
        if element.category == 'Table':
            current_chunk += element.html + '\n'
        else:
            current_chunk += element.text + '\n'
    
    if current_chunk:
        chunks.append(current_chunk)

    return chunks

In [95]:
# Chunk the elements
chunks = chunk_by_previous_title(test_cleaned_elements)

In [96]:
# Print the chunks
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1} \n")
    print(chunk)
    print("\n\n" + "End of Chunk" + "-"*80 + "\n\n")

Chunk 1 

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are unco

In [99]:
unstructured_chunking = chunk_by_title(test_cleaned_elements)

for i in unstructured_chunking:
    print(i.to_dict())

print(len(unstructured_chunking))

{'type': 'CompositeElement', 'element_id': 'c8893f10ac07fe6e7b3f13a093dc899f', 'text': 'CROSSLIN TRANSIT SOLUTIONS\n\nOct 20, 2021 \n\nCyber Security – Change Control and Configuration Management Process', 'metadata': {'file_directory': 'c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs', 'filename': '5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf', 'filetype': 'application/pdf', 'languages': ['en'], 'last_modified': '2023-03-28T16:34:32', 'page_number': 1, 'orig_elements': 'eJztlEtr4zAUhf+K8Lqx9fAzu5LAEOgkpc4wi04xsiQnAlsyktxpKP3vIzkpDFMoQ6G7ghe+R/dcWUcfvn+ORC8GoVwjebQEUYsIraqOtThlkKKiEAVhbV4hwipKyi66AtEgHOXUUd//HDGtDZeKOmHnuqcnPbnmKOTh6LyCMYTec5F/S+6OXkXFrI5aKhd89/eowjG5AgTBOH+4Aq91RnCchrpIqxi/rc/9XojsyToxhDPcyifR1yNlInrxC53sRcOlEcxpcwoNbJn8sMLYZE1VslNibeSjSNaaTSEJm6yMtraX6ilZ1V610WWKooMI/gxCuPDPz2/rRVreXi8QxBAswOpI1UGAlVbO6B5QxcN7Jw+ToU5qBb5TRQ9z3ODWaCasjUfevY53p3EeT8exl2x2JJfl3g+evHXOKhIqephF65pBc9lJMV8ehpgsIFngco/yJUmXBAfz6I2NmoZWmBB9yMSJp3A50epuV9c3